### ISSUE TITLE
add turn-on MathJax equation numbering for math environments

### ISSUE THREAD

**Issue by eeholmes** [quarto-dev/quarto-cli#4136]
By default, MathJax does not have equation numbers when the LaTeX math environments are used. This:
```
---
title: "test6"
format: html
---

\begin{equation}\label{eq-a}
a
\end{equation}
```
produces this:
<img width="757" alt="image" src="https://user-images.githubusercontent.com/2545978/215232056-a2ab5e88-27c5-41b1-a348-803d96c652cb.png">

In PDF, it will produce the equation numbers.

The reason is that the default behavior for MathJax is to not show the equation numbers. This causes endless confusion for those who use the math environment, which is used for more complex math.

It'd be great if Quarto turned on equation numbering in math environments by default so that html and PDF output behaves the same. All that is needed is to have this script in the header if the equation engine is mathjax:
```
<script>
MathJax = { 
    tex: { 
        tags: 'ams'  // should be 'ams', 'none', or 'all' }
     };
</script>
```
Obviously you can turn this on manually with
```
---
title: "test5"
format:
    html:
      html-math-method: mathjax
      include-in-header:
        - text: |
            <script>
              MathJax = { 
                tex: { 
                  tags: 'ams'  // should be 'ams', 'none', or 'all' }
                };
            </script>
---
```
but most users spend hours and hours trying to debug lack of numbers until they stumble upon the right solution. What is worse, is that the solution depends on the MathJax version. The above is the method for the version being used by Quarto (by default).

I don't think there is any reason to have the numbering turned off. Anyone using the LaTeX math environment is likely to be a LaTeX user and we have a standard methods for selectively turning off numbers on an equation if that were wanted.

**Comment by cscheid** [comment]
Thanks for the feedback! I think what you're saying makes sense. We need to check how these interact with our crossref systems, but I'll take a look at that as we're overhauling crossrefs internals in any case

**Comment by eeholmes** [comment]
Hi @cscheid Feel free to reach out to me for any feedback on the math side of the crossref system. I am heavily invested in the LateX-Quarto interface (with quarto_titlepage and other mathy templates). I have years of experience in complex math in the Rmd ecosystem. Did this heavily in bookdown but since migrated to Quarto. As an applied mathematician, I am familiar with how those who do complex math would expect math cross-referencing to work.  

**Re the MathJax options**

I am guessing it is probably better to set math options in the metadata rather than hard-coding anything in header.
```
html-math-method:
  method: mathjax
  number-equations: true
  add-to-equation: section
```
For example adding section numbers to equations numbers is common but the [MathJax setting](https://docs.mathjax.org/en/latest/input/tex/extensions/tagformat.html#example-section-numbering) to add section numbers to equation numbers is complex. Note in LaTeX it's `\numberwithin{equation}{section}` but doesn't work in html output, only PDF.

**Re, cross-referencing in math equations**

In the spirit of not re-inventing the wheel, LaTeX (which MathJax and ilk know) already has a system for complex math (latex math environments) and how to cross-ref. Anyone who puts this kind of math in their documents knows and uses the LaTeX/AMS syntax for math equations. The `$$ $$ {#eq-num}` is great actually but only for simple stuff, after that one goes to `\begin{math-env} ...`.  With MathJax (and other LaTeX/AMS to html JS libraries), I have found that I can just write normal LaTeX/AMS syntax and everything works same in html and pdf. It's beautiful. The only common thing that MathJax doesn't do is the `subequations` environment, but maybe that is kind of niche, or at least there hasn't been enough community effort to robustly solve it.

I think it is key to have have LaTeX/AMS users give feedback in order not to get "painted in a corner" by doing a math cross-ref system that will not allow the core cross-ref syntax that every math user will want. Like, numbering of sub-equations in a math environment and the different ways one does that in the LaTeX/AMS syntax (there really only 2). Even if your first system can't do that in the beginning, I bet there are ways to design things so that that will be a doable update in the future versus having to rebuild the whole system from the ground up.

BTW, currently Quarto works for fine for complex math and equation cross-refs. I just have to be careful to never use `$$ $$ {#eq-math}` (and `@eq-math`) and only use LaTeX math environments (no `$$ $$`) and use LaTeX eqn cross-ref syntax. So like `\label{eq-num}` and `\ref{eq-num}`. Sadly I have to avoid `\eqref{eq-num}` but that is a Pandoc problem with only recognizing newly defined LaTeX commands in `$ $` environment (grrr). It is weird to have to use `@` for figs and tables but `\ref{}` for equations but it's not that a big of a deal.

Cheers!

**Comment by cscheid** [comment]
I will do, and thank you for the long, thoughtful followup.

**Comment by eeholmes** [comment]
Adding a couple more notes (more for myself). MathJax fully supports amsmath so if you are having trouble with your cross-refs showing up, double check that your amsmath syntax is correct.  *Note if you happen to land on this thread and you are working with RMarkdown, be aware that in May 2023, RMarkdown still uses MathJax 2.7 (from 2017!) and the configuration syntax for that is completely different than MathJax 3+.*

Example mathjax.html (to include in header for html output)

If you need your equations numbered by section, so like 1.2 and 3.1:
```
<!-- For MathJax 3+ Eqns nums with section numbers -->
<script>
MathJax = {
  section: 1,
  tex: {
    tags: 'ams',
    tagformat: {
      number: (n) => MathJax.config.section + '.' + n
    }
  }
};
</script>
```

If you want your `$$ ... $$` equations numbered, change `ams` to `all` however note that you cannot cross-ref your equations if you do this. `$$a+b \label{eq:a}$$` with this  `\ref{eq:a}` won't work.
```
<script>
MathJax = {
  tex: {
    tags: 'all',
    }
};
</script>
```

If you run into trouble with MathJax gettting "confused" as to what section you are in, then this is more robust:
```
<script>
MathJax = {
  section: 1,
  tex: {
    tags: 'ams',
    tagformat: {
      number: (n) => MathJax.config.section + '.' + n,
      id: (tag) => 'eqn-id:' + tag
    }
  },
  startup: {
    ready() {
      MathJax.startup.defaultReady();
      MathJax.startup.input[0].preFilters.add(({math}) => {
        if (math.inputData.recompile) {
          MathJax.config.section = math.inputData.recompile.section;
        }
      });
      MathJax.startup.input[0].postFilters.add(({math}) => {
        if (math.inputData.recompile) {
          math.inputData.recompile.section = MathJax.config.section;
        }
      });
    }
  }
};
</script>
``` 

**Quarto users ignore this: This is for the poor Rmd folks who might land on this thread in desperation.** As of May 2023, the default MathJax js file that getting imported for Rmd is MathJax 2.7. You need to use a more recent version of MathJax to get proper support for equation numbering and cross-refs. Use this in your yaml:
```
output: 
  html_document: 
    mathjax: "https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js"
    includes:
      in_header: mathjax.html
```
with the mathjax.html files above depending on what you need.

**Comment by ute** [comment]
Thank you so much for sharing, @eeholmes ! The newer mathjax also supports to change equation numbers to custom tags, and `\eqref` is equally supported. That makes migration from LaTeX to quarto so so much easier :-) (in my case: tons of course notes with complex mathematics)
I will throw together a filter that also takes care of markdown $$-equations.

My addition here to the wish list for quarto crossref: a mechanism for **\tag**ging. For equations, one could keep LaTeX syntax, that is what presumably most users of such a feature already are familiar with.

**Comment by ute** [comment]
Here is a first version of the filter:
https://github.com/ute/mathjax3eqno
:-)
Although I do not load a specific mathjax version, it works on my machine with quarto 1.3.353. I can see that quarto loads MathJax3 👍 
I have not tried it with earlier versions

**Comment by cscheid** [comment]
See also #2275

**Comment by dhodge180** [comment]
Thanks @ute your extension is great, I've been searching around for something that allows equation numbering to work out-of-the-box.
I don't have any great need for the numbering my chapter etc... but the lua script function (Para) is great, It replaces
```
$$
  stuff
$$ {#eq-xx}
```
with
```
\begin{equation}
  stuff \label{eq-xx}
\end{equation}
```

thus allowing usage of Markdown $$ syntax when that's all I need. But for more complex numbering I can still use LaTeX syntax via a math environment. The only drawback is we don't have a filter for `\[ .... \]` environments. So we just have to use either `$$...$$` or specific math environments like align, equation, etc..

For anyone just finding this thread, you only need to include the mathjax1.html file (in the header) to enable to ams labelling, and then use Ute's nice Para function (see the filter [mathjax3eqno.lua](https://github.com/ute/mathjax3eqno/blob/main/_extensions/mathjax3eqno/mathjax3eqno.lua)).

In my writing I'd like to be able to use Markdown format most of the time, but fallback on LaTeX/MathJax for things that Markdown cannot do (like number some lines in an align, but not others). While I wait for Quarto to get its act together and support numbering of equations in a more complete way. Is there any sign that Quarto will be updated to allow for richer equation numbering?

I'll be checking back to see if you have any success with making it compatible with Quarto books (or if Quarto fixes it all before then).

**Comment by ute** [comment]
@dhodge180 , there are two problems with books, due to the fact that chapters are rendered kind of independently:

a. mathjax does not know the chapter number, and uses chapter sections as prefix for equation numbers,

b.  cross-references between chapters cannot be resolved by mathjax

Problem a. could be fixed with tweaking mathjax, but b. amounts to rewriting the crossreferencing mechanism. I believe the quarto team will come up with a solution soon-ish.

I also need this equation numbering in a booj right now, and currently work around by setting latex tags manually. This works with mathjax.

**Comment by dhodge180** [comment]
> @dhodge180 , there are two problems with books, due to the fact that chapters are rendered kind of independently:
> 
> a. mathjax does not know the chapter number, and uses chapter sections as prefix for equation numbers,
> 
> b. cross-references between chapters cannot be resolved by mathjax
> 
> Problem a. could be fixed with tweaking mathjax, but b. amounts to rewriting the crossreferencing mechanism (which also would solve a). I believe the quarto team will come up with a solution soon-ish.
> 
> I also need this equation numbering in a book right now, and currently work around by setting latex tags manually. This works with mathjax.

Thanks! I've had a play with your code to see if I can help with a. but I'm not good enough with Lua I don't think. I can't work out yet why the `$$...$$ {#eq-1}` equations already have hard-coded `\qquad(1.1)` style tags. Probably pandoc processing before your filter gets hold of them.

For the meantime for a book I think you're saying the answer is probably to use (display) `$$...$$ {#eq-1}` and (inline) `$...$` as per Quarto standard, and if I ever need to label multiple lines of an equation to perhaps give them non-numeric names via `\tag{$\dagger_1$}`, `\tag{$\dagger_2$}` etc... this way when Quarto does provide a way to support better numbering I won't have to change all my code, only these special cases?

**Comment by ute** [comment]
@dhodge180 , there are two problems with books, due to the fact that chapters are rendered kind of independently:

a. mathjax does not know the chapter number, and uses chapter sections as prefix for equation numbers,

b.  cross-references between chapters cannot be resolved by mathjax

Problem a. could be fixed with tweaking mathjax, but b. amounts to rewriting the crossreferencing mechanism. I believe the quarto team will come up with a solution soon-ish.

I also need this equation numbering in a book right now, and currently work around by setting latex tags manually. This works with mathjax.

**Comment by ute** [comment]
@dhodge180, I have updated my extension with a workaround for books. It is still not possible to cross reference between chapters, but it works within a chapter now. You have to provide the chapter number manually with the first equation, then all equation numbers im that chapter are prefixed with chapter number.
Crossref between chapters works only in pdf. It would be quite epic to implement this for html books.

**Comment by ute** [comment]
@dhodge180, I have updated my extension with a workaround for books. It is still not possible to cross reference between chapters, but it works within a chapter now. You have to provide the chapter number manually with the first equation, then all equation numbers im that chapter are prefixed with chapter number.
Crossref between chapters works only in pdf. It would be quite epic to implement this for html books.

**Comment by Ar-Kan** [comment]
Hey all,

I come up with a solution that works kinda well on the HTML version, basically it looks for elements that already exists on the DOM (only tested in the book format).

It looks for a span element with the class `chapter-number` inside the active chapter in the sidebar, if found, it will use the text content
as the chapter number, otherwise, it will default to 0.

Here is the code:

```html
<script>
  window.MathJax = {
    loader: { load: ["[tex]/tagformat"] },
    tex: {
      tags: "ams",
      packages: { "[+]": ["tagformat", "sections"] },
      tagformat: {
        number: (n) => {
          if (n === "") return "";
          // FIXME: The're still no answer for unnumbered chapters sections,
          //        for now, we should avoid putting equations there :P

          // NOTE: Get the chapter number element from the sidebar (the highlighted one),
          //       if found the numbering would be, for example, "0.1", "1.1", etc.
          const chapterNumber = document.querySelector(
            "#quarto-sidebar a.sidebar-link.active span.chapter-number"
          );
          // NOTE: Get the value, defaults to 0 if not found (unnumbered chapters)
          chapter = chapterNumber ? chapterNumber.textContent : 0;
          return `${chapter}.${n}`;
        },
      },
    },
  };
</script>
```

An idea for numberings like "1.1.1", "A.1.1", etc. is to use the TOC, if the chapter is not found, it will default to "0.1", "1.1", etc.

Something like:

```js
const toc = document.querySelector("#TOC li:has(a.active)");
const chapterToc = toc
  ? toc.getElementsByClassName("header-section-number")
  : undefined;

let chapter;
if (chapterToc && chapterToc.length) {
  chapter = chapterToc[0].textContent;
}
```

But that will only work if the're only one section in the page.
A better solution would be to look for the closest header element, but that will require to know where the equation is on the DOM, I'm not sure how to do that with MathJax.
